### ISSUE TITLE
Calculus: gradients (intuition, not proofs)

### ISSUE THREAD

**Issue by zahrahanifehlou** [zahrahanifehlou/protein-function-predict#7]

**Comment by zahrahanifehlou** [comment]
### Mini-Batch Gradients

In machine learning, we train models by updating parameters (weights and biases) to minimize a loss function using gradients. How we compute these gradients depends on how much data we use at once.

**Batch Gradient Descent**
Use all training samples to compute the gradient.
Formula (Linear Regression example):

<img width="221" height="73" alt="Image" src="https://github.com/user-attachments/assets/e6fc940e-4e1e-46f0-a0b8-bac42dc3b04a" />

where 
N = total number of samples.
Pros: Accurate gradient.
Cons: Slow if dataset is large.

**Stochastic Gradient Descent (SGD)**
Use one sample at a time.
Formula:
<img width="164" height="77" alt="Image" src="https://github.com/user-attachments/assets/b42a3fe0-00b4-4c85-9992-ccdc52fc5d8a" />

Pros: Fast, can escape local minima.
Cons: Noisy gradient → can be unstable.

**Mini-Batch Gradient Descent** 
Use a small subset (“mini-batch”) of samples to compute the gradient.
Formula: 

<img width="225" height="90" alt="Image" src="https://github.com/user-attachments/assets/8aff2b26-62c7-403e-8889-70d81542302b" />

where 
B = mini-batch size.

✅ It’s a balance between batch and stochastic gradient descent.



Reduces noise compared to SGD.
Faster than full-batch.
Works well on GPUs for parallel computation.
